Add authorization support for macOS and an example application.

msarda opened the pull request:
This PR adds supports to do an authorization request following the OAuth 2 for Installed Applications flow (https://developers.google.com/identity/protocols/OAuth2InstalledApp) for macOS.
The main differences between iOS and macOS:
* macOS opens the default browser for the authorization flow (it cannot use an embedded Safari View Controller)
* macOS requires a client secret when exchanging the authorization code for an refresh token.

Just like the iOS Example, it uses a Custom URI Scheme that allows the browser to open back the example application once the user authorizes the access. The code of the AppAuthExampleViewController. app is a copy-paste with same small modifications of the iOS example one.

Note that the Example-Mac xcode project uses pods to configure the

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/openid/appauth-ios/35)
<!-- Reviewable:end -->

msarda commented:
Review status: 0 of 28 files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KMOI3N3u4PAWv1A2Nbk:-KMOI3N3u4PAWv1A2Nbl:-815055961) (no related file):*
I fixed the test breakage. Please take a another look.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/35)*
<!-- Sent from Reviewable.io -->

WilliamDenniss commented:
Review status: 0 of 28 files reviewed at latest revision, 4 unresolved discussions.

---

*[Example-Mac/Source/AppAuthExampleViewController.h, line 4 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN2sIPE7uJUO7C0b5of:-KN2sIPE7uJUO7C0b5og:-1855650777) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.h#L4)):*
> ```objc
>     @brief AppAuth macOS SDK Example
>     @copyright
>         Copyright 2015 Google Inc. All Rights Reserved.
> ```

nit: 2016 for new code

---

*[Source/OIDAuthorizationRequest.h, line 71 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN2rYllnyer1Pt3mQp4:-KN2rYllnyer1Pt3mQp5:-1270316407) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Source/OIDAuthorizationRequest.h#L71)):*
> ```objc
>     @remarks client_secret
>     @discussion The client secret is not needed on iOS, but is mandatory when exchanging the
>         authorization code for an access token on macOS.
> ```

This iOS vs macOS detail is true for Google's AS but not globally. Suggest replacing this text with a simple factual explanation of what this param is.

---

*[Source/OIDAuthorizationRequest.h, line 155 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN2rs_aXjWH7T3pIuy-:-KN2rs_aXjWH7T3pIuy0:2002146749) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Source/OIDAuthorizationRequest.h#L155)):*
> ```objc
>         automatically.
>  */
> - (nullable instancetype)
> ```

does this need to be nullable?  that was removed in https://github.com/openid/AppAuth-iOS/pull/33

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/35)*
<!-- Sent from Reviewable.io -->

StevenEWright commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 25 of 26 files at r1, 2 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 35 unresolved discussions.

---

*[Example-Mac/README.md, line 33 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN2zzJRlcMUa9eKdl_g:-KN2zzJRlcMUa9eKdl_h:1483934896) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/README.md#L33)):*
> id

ID for consistency.

---

*[Example-Mac/Source/AppAuthExampleViewController.h, line 31 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN3-VI9HOCc_N-XcjeE:-KN3-VIAFmSNxPq7eR7g:-755392918) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.h#L31)):*
> ```objc
> @interface AppAuthExampleViewController : NSViewController
> 
> @property IBOutlet NSButton *authAutoButton;
> ```

Should these IBOutlets be marked (nullable) and documented?

I realize it's a slippery slope, but I don't have a super strong opinion about making sure every single property is documented in the examples... though I'd default to documenting them all if in doubt. William, WDYT? It may be /even more important/ for the examples to be clear and documented than the actual code. I'm not sure.

---

*[Example-Mac/Source/AppAuthExampleViewController.h, line 38 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN30yVBwmz8P1S2BqIz:-KN30yVBwmz8P1S2BqJ-:-2076779471) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.h#L38)):*
> @property (nonatomic, weak) AppDelegate *appDelegate;

I'd definitely document this since it's not clear why you have a property for it. Why not just 

---

*[Example-Mac/Source/AppAuthExampleViewController.h, line 44 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN30aY6PnLbool9zRYb:-KN30aY6PnLbool9zRYc:-2084977817) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.h#L44)):*
> strong,

If it's readonly I don't think we need the memory management semantics to be declared here.

---

*[Example-Mac/Source/AppAuthExampleViewController.h, line 50 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN30looX3OXwB6zPT-s:-KN30looX3OXwB6zPT-t:58341081) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.h#L50)):*
> (nullable id)sender

The sender parameter is optional. If you're not using it, I'd recommend removing it for clarity. Though I don't feel strongly about it.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 4 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN32-QRHVmdHZfdjSS0:-KN32-QRHVmdHZfdjSS1:-631421774) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L4)):*
> 2015

And elsewhere.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 37 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN328be_RIjz2mz0nhe:-KN328be_RIjz2mz0nhf:737417410) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L37)):*
> ```objc
>  */
> static NSString *const kClientID =
>     @"YOUR_CLIENT.apps.googleusercontent.com";
> ```

Doesn't fit on one line?

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 62 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN32JUo4XY9gSE6-rPg:-KN32JUo4XY9gSE6-rPh:-1928694086) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L62)):*
> ```objc
> 
> @interface AppAuthExampleViewController () <OIDAuthStateChangeDelegate, OIDAuthStateErrorDelegate>
> - (void)setAuthState:(nullable OIDAuthState *)authState;
> ```

We shouldn't have to forward declare Objective-C methods anymore. Why is this here?

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 80 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN32Z1X4ymlIeS4vM1S:-KN32Z1X4ymlIeS4vM1T:-2096415017) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L80)):*
> NSAssert

Really nice. Super like. <3

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 81 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN32UUoGBYjsIkfXzip:-KN32UUoGBYjsIkfXziq:1483934896) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L81)):*
> id

ID for consistency.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 135 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN32vyjXIc3TdybVFEf:-KN32vykRYMA6TJBrbIA:1252103829) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L135)):*
> ```objc
> 
> - (void)setAuthState:(nullable OIDAuthState *)authState {
>   _authState = authState;
> ```

Consider adding:

if (_authState == authState) {
  return;
}

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 145 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN338DM9uQ5Z6JtXBIF:-KN338DM9uQ5Z6JtXBIG:217890529) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L145)):*
> ```objc
> - (void)updateUI {
>   _userinfoButton.enabled = [_authState isAuthorized];
>   _clearAuthStateButton.enabled = (_authState != nil);
> ```

Unneeded parentheses.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 146 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN33CROpNSGjtB0q9pt:-KN33CROpNSGjtB0q9pu:-485711229) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L146)):*
> ```objc
>   _userinfoButton.enabled = [_authState isAuthorized];
>   _clearAuthStateButton.enabled = (_authState != nil);
>   _codeExchangeButton.enabled = (_authState.lastAuthorizationResponse.authorizationCode
> ```

Same.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 273 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN33dcYUtdTxMa87YKm:-KN33dcYUtdTxMa87YKn:-993187760) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L273)):*
> ```objc
>   [OIDAuthorizationService performTokenRequest:tokenExchangeRequest
>                                       callback:^(OIDTokenResponse *_Nullable tokenResponse,
>                                                NSError *_Nullable error) {
> ```

Vertical alignment.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 312 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN34BCwS4lDFN2GF7CS:-KN34BCwS4lDFN2GF7CT:-1895307405) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L312)):*
> ```objc
> 
>     // log whether a token refresh occurred
>     if (currentAccessToken != accessToken) {
> ```

I don't think reference equality is what you want here.

if (![currentAccessToken isEqual:accessToken]) {

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 332 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN35q-B-kQm-szr7s3l:-KN35q-B-kQm-szr7s3m:1076710959) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L332)):*
> NSURLSessionDataTask

FYI @William; min OS X v10.9 (LGTM)

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 338 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN34SRwLHT95wQR2wuc:-KN34SRwLHT95wQR2wud:431705615) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L338)):*
> ```objc
>                                        NSError *_Nullable error) {
>       dispatch_async(dispatch_get_main_queue(), ^() {
> ```

Inconsistent empty line at beginning of block.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 340 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN36ihS8ChVSFjcKGGl:-KN36ihS8ChVSFjcKGGm:-437115837) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L340)):*
> error];

Perhaps it'd be good to specifically check for if (error) and if (![response isKindOfClass:...]) separately? 

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 347 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN36xg0hS0sO20WxX2P:-KN36xg0hS0sO20WxX2Q:-1489187660) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L347)):*
> :&jsonError

You never check if JSONError is non-nil after this method returns. I'd say either check it and take appropriate action, or don't pass the ref.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 349 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN34hVIMElyGj607rtO:-KN34hVIMElyGj607rtP:-1258063750) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L349)):*
> if (httpResponse.statusCode != 200) {

@William, this is very strict. Need it be this strict?

---

*[Example-Mac/Source/AppDelegate.h, line 22 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN37Oe_FcbEVRO6bfRH:-KN37Oe_FcbEVRO6bfRI:566601234) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppDelegate.h#L22)):*
> ```objc
> 
> @protocol OIDAuthorizationFlowSession;
> ```

NS_ASSUME_NONNULL_BEGIN ... 

---

*[Example-Mac/Source/AppDelegate.m, line 4 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN381eUIVyJ7Grxn1fE:-KN381eUIVyJ7Grxn1fF:1280947244) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppDelegate.m#L4)):*
> 2015

2016

---

*[Example-Mac/Source/AppDelegate.m, line 21 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN37WahnHW6sPZwQIwC:-KN37WahnHW6sPZwQIwD:1035070940) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppDelegate.m#L21)):*
> ```objc
> #import "AppDelegate.h"
> 
> #import "AppAuth.h"
> ```

Shouldn't this be #import <AppAuth/AppAuth.h> ? (If we are assuming we are consuming the framework?)

---

*[Example-Mac/Source/AppDelegate.m, line 25 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN37oe77XMcCldGOc7d:-KN37oe77XMcCldGOc7e:1131334314) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppDelegate.m#L25)):*
> ```objc
> 
> @interface AppDelegate ()
> @property (weak) IBOutlet NSWindow *window;
> ```

IBOutlets on macOS are assumed to be weak per style guide.

---

*[Example-Mac/Source/Assets.xcassets/AppIcon.appiconset/Contents.json, line 1 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN38FbNE1sMBjfDxN1q:-KN38FbNE1sMBjfDxN1r:-752217768) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/Assets.xcassets/AppIcon.appiconset/Contents.json#L1)):*
> ```JSON
> {
> ```

But we don't actually have any app icons :(

---

*[Example-Mac/Source/Base.lproj/MainMenu.xib, line 27 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN38QfEOc21JfUyzyG1:-KN38QfEOc21JfUyzyG2:241169230) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/Base.lproj/MainMenu.xib#L27)):*
> ```XML
>                     <menu key="submenu" title="Example-Mac" systemMenu="apple" id="uQy-DD-JDr">
>                         <items>
>                             <menuItem title="About Example-Mac" id="5kV-Vb-QxS">
> ```

Nit; Can we tweak the menu menu nib so the appropriate app descriptors are used (Like "AppAuth Example" or something?)

---

*[Example-Mac/Source/Base.lproj/MainMenu.xib, line 34 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN38aIdfOKfr39gPiiB:-KN38aIdfOKfr39gPiiC:-591360682) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/Base.lproj/MainMenu.xib#L34)):*
> ```XML
>                             </menuItem>
>                             <menuItem isSeparatorItem="YES" id="VOq-y0-SEH"/>
>                             <menuItem title="Preferences…" keyEquivalent="," id="BOF-NM-1cW"/>
> ```

And let's get rid of the unused ones like "Preferences..."

---

*[Source/AppAuth.h, line 38 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN38lXnn1P1RW52nDDi:-KN38lXohRaJp7sxdST1:-196846373) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Source/AppAuth.h#L38)):*
> ```objc
> #import "OIDAuthState+IOS.h"
> #import "OIDAuthorizationService+IOS.h"
> #else
> ```

Consider:

#elif TARGET_OS_MAC
...
#else
#error "Platform Undefined; please specify TARGET_OS_MAC or TARGET_OS_IPHONE"
#endif

---

*[Source/OIDAuthorizationRequest.h, line 155 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN2rs_aXjWH7T3pIuy-:-KN39ElxtVKO3nSphOVA:5172523) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Source/OIDAuthorizationRequest.h#L155)):*
<details><summary><i>Previously, WilliamDenniss (William Denniss) wrote…</i></summary>
> does this need to be nullable?  that was removed in https://github.com/openid/AppAuth-iOS/pull/33
</details>
+1

---

*[Source/OIDAuthorizationRequest.h, line 176 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN39I1_-h5OnIJH9w2C:-KN39I1_-h5OnIJH9w2D:677068555) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Source/OIDAuthorizationRequest.h#L176)):*
> ```objc
>         automatically.
>  */
> - (nullable instancetype)initWithConfiguration:(OIDServiceConfiguration *)configuration
> ```

Same

---

*[Source/OIDAuthorizationRequest.h, line 178 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN39L-2q_Tz2WStD1x9:-KN39L-2q_Tz2WStD1xA:1039040779) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Source/OIDAuthorizationRequest.h#L178)):*
> ```objc
> - (nullable instancetype)initWithConfiguration:(OIDServiceConfiguration *)configuration
>                 clientId:(NSString *)clientID
>             clientSecret:(nullable NSString *)clientSecret
> ```

Consider moving nullable parameters to end of list. No strong feeling if ordering makes more sense as is.

---

*[Source/macOS/OIDAuthorizationUICoordinatorMac.h, line 27 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN3A1A5cMJxZa6IPmy9:-KN3A1A6AkH6Fb0EAFbq:849694973) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Source/macOS/OIDAuthorizationUICoordinatorMac.h#L27)):*
> ```objc
>         present an authorization request.
>  */
> @interface OIDAuthorizationUICoordinatorMac : NSObject<OIDAuthorizationUICoordinator>
> ```

In other files you've added a space between the base class and the protocol list.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/35#-:-KN3Ahiq7N_TUNKQMqRr:980077423)*
<!-- Sent from Reviewable.io -->

msarda commented:
Review status: 18 of 30 files reviewed at latest revision, 34 unresolved discussions.

---

*[Example-Mac/README.md, line 33 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN2zzJRlcMUa9eKdl_g:-KN6Pf_vBogbsMQnVcpI:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/README.md#L33)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> > id
> 
> ID for consistency.
</details>
Done.

---

*[Example-Mac/Source/AppAuthExampleViewController.h, line 4 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN2sIPE7uJUO7C0b5of:-KN6PnxnzlNQdDySmYee:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.h#L4)):*
<details><summary><i>Previously, WilliamDenniss (William Denniss) wrote…</i></summary>
> nit: 2016 for new code
</details>
Done.

---

*[Example-Mac/Source/AppAuthExampleViewController.h, line 31 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN3-VI9HOCc_N-XcjeE:-KN6TWJ25lDr42-0lOt1:-531515206) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.h#L31)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Should these IBOutlets be marked (nullable) and documented?
> 
> I realize it's a slippery slope, but I don't have a super strong opinion about making sure every single property is documented in the examples... though I'd default to documenting them all if in doubt. William, WDYT? It may be /even more important/ for the examples to be clear and documented than the actual code. I'm not sure.
</details>
Done for the nullable.

I would not document these IBOutlets. Having them marked as IBOutlets makes them clear (anybody that wonders what these properties are can just open the XIB file.)

---

*[Example-Mac/Source/AppAuthExampleViewController.h, line 38 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN30yVBwmz8P1S2BqIz:-KN6xgxSdE1JxWewvk6z:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.h#L38)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> > @property (nonatomic, weak) AppDelegate *appDelegate;
> 
> I'd definitely document this since it's not clear why you have a property for it. Why not just 
</details>
Done.

---

*[Example-Mac/Source/AppAuthExampleViewController.h, line 44 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN30aY6PnLbool9zRYb:-KN6UpucmqC5XZgYEICF:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.h#L44)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> > strong,
> 
> If it's readonly I don't think we need the memory management semantics to be declared here.
</details>
Done.

---

*[Example-Mac/Source/AppAuthExampleViewController.h, line 50 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN30looX3OXwB6zPT-s:-KN6UtZpXs6qZxPBDB_Q:-1341531926) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.h#L50)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> > (nullable id)sender
> 
> The sender parameter is optional. If you're not using it, I'd recommend removing it for clarity. Though I don't feel strongly about it.
</details>
I'm keeping it to match the iOS example.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 4 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN32-QRHVmdHZfdjSS0:-KN6Q4tOGJA7r9jhT0PX:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L4)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> > 2015
> 
> And elsewhere.
</details>
Done.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 37 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN328be_RIjz2mz0nhe:-KN6V6YFkwAa_CKSJGQP:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L37)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Doesn't fit on one line?
</details>
Done.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 62 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN32JUo4XY9gSE6-rPg:-KN6VLn9scbVIqUAHpDp:159871161) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L62)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> We shouldn't have to forward declare Objective-C methods anymore. Why is this here?
</details>
Done (copy-paste is evil).

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 81 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN32UUoGBYjsIkfXzip:-KN6VlYct7jeEEylIQoN:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L81)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> > id
> 
> ID for consistency.
</details>
Done.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 135 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN32vyjXIc3TdybVFEf:-KN6VvNMd1f_f27APqPc:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L135)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Consider adding:
> 
> if (_authState == authState) {
>   return;
> }
> 
</details>
Done.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 145 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN338DM9uQ5Z6JtXBIF:-KN6xjd0pr6am8Gs4iIe:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L145)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Unneeded parentheses.
</details>
Done.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 146 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN33CROpNSGjtB0q9pt:-KN6W0yfdYn19z_1RfWN:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L146)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Same.
</details>
Done.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 273 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN33dcYUtdTxMa87YKm:-KN6WF2qHGQ9VJa_UzNA:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L273)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Vertical alignment.
</details>
Done.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 312 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN34BCwS4lDFN2GF7CS:-KN6X_-EdCKKybS-ITQf:1056692906) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L312)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> I don't think reference equality is what you want here.
> 
> if (![currentAccessToken isEqual:accessToken]) {
</details>
I think the current test is valid as the property _authState.lastTokenResponse.accessToken is readonly and not copied. So comparing the 2 references works. I changed the test as testing the values is actually better.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 338 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN34SRwLHT95wQR2wuc:-KN6Z3aJN8CnEkTnfNO8:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L338)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Inconsistent empty line at beginning of block.
</details>
Done.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 340 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN36ihS8ChVSFjcKGGl:-KN6ZerHYv2MW4bgY5ur:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L340)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> > error];
> 
> Perhaps it'd be good to specifically check for if (error) and if (![response isKindOfClass:...]) separately? 
</details>
Done.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 347 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN36xg0hS0sO20WxX2P:-KN6_0Mw7px6O2BvWX10:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L347)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> > :&jsonError
> 
> You never check if JSONError is non-nil after this method returns. I'd say either check it and take appropriate action, or don't pass the ref.
</details>
Done.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 349 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN34hVIMElyGj607rtO:-KN6_1ZG34igJL7mdBDS:1422477630) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppAuthExampleViewController.m#L349)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> > if (httpResponse.statusCode != 200) {
> 
> @William, this is very strict. Need it be this strict?
</details>
In my experience, only 200 is a success. We should not consider success if the status is something else than 200.

---

*[Example-Mac/Source/AppDelegate.h, line 22 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN37Oe_FcbEVRO6bfRH:-KN6_Tvessi4qVWA3uYv:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppDelegate.h#L22)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> NS_ASSUME_NONNULL_BEGIN ... 
</details>
Done.

---

*[Example-Mac/Source/AppDelegate.m, line 4 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN381eUIVyJ7Grxn1fE:-KN6_Ukele-vp1d0vrav:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppDelegate.m#L4)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> > 2015
> 
> 2016
</details>
Done.

---

*[Example-Mac/Source/AppDelegate.m, line 21 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN37WahnHW6sPZwQIwC:-KN6xl0Hqcm-T_gd94Rj:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppDelegate.m#L21)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Shouldn't this be #import <AppAuth/AppAuth.h> ? (If we are assuming we are consuming the framework?)
</details>
Done.

---

*[Example-Mac/Source/AppDelegate.m, line 25 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN37oe77XMcCldGOc7d:-KN6_iAsKaV36Odp7f8K:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/AppDelegate.m#L25)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> IBOutlets on macOS are assumed to be weak per style guide.
</details>
Done.

---

*[Example-Mac/Source/Assets.xcassets/AppIcon.appiconset/Contents.json, line 1 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN38FbNE1sMBjfDxN1q:-KN6dB3fQ1tkEfFNqQBq:-1085003166) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/Assets.xcassets/AppIcon.appiconset/Contents.json#L1)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> But we don't actually have any app icons :(
</details>
This is the default structure of the app created by xcode. I think it is fine if we do not add an icon for the example app.

I prefer keeping this file to avoid fiddling with the xcodeproj config.

---

*[Example-Mac/Source/Base.lproj/MainMenu.xib, line 27 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN38QfEOc21JfUyzyG1:-KN6fBRvqnTsCUTzdq0i:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/Base.lproj/MainMenu.xib#L27)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Nit; Can we tweak the menu menu nib so the appropriate app descriptors are used (Like "AppAuth Example" or something?)
</details>
Done.

---

*[Example-Mac/Source/Base.lproj/MainMenu.xib, line 34 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN38aIdfOKfr39gPiiB:-KN6fBnk6rGfYYER8-2a:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Example-Mac/Source/Base.lproj/MainMenu.xib#L34)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> And let's get rid of the unused ones like "Preferences..."
</details>
Done.

---

*[Source/AppAuth.h, line 38 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN38lXnn1P1RW52nDDi:-KN6_vDuV_Kuw7w0Vk6C:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Source/AppAuth.h#L38)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Consider:
> 
> #elif TARGET_OS_MAC
> ...
> #else
> #error "Platform Undefined; please specify TARGET_OS_MAC or TARGET_OS_IPHONE"
> #endif
> 
</details>
Done.

---

*[Source/OIDAuthorizationRequest.h, line 71 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN2rYllnyer1Pt3mQp4:-KN6bWk3-vpUCHGsW-WH:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Source/OIDAuthorizationRequest.h#L71)):*
<details><summary><i>Previously, WilliamDenniss (William Denniss) wrote…</i></summary>
> This iOS vs macOS detail is true for Google's AS but not globally. Suggest replacing this text with a simple factual explanation of what this param is.
</details>
Done.

---

*[Source/OIDAuthorizationRequest.h, line 155 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN2rs_aXjWH7T3pIuy-:-KN6xezn_l0V-Jvl7X0_:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Source/OIDAuthorizationRequest.h#L155)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> +1
</details>
Done.

---

*[Source/OIDAuthorizationRequest.h, line 176 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN39I1_-h5OnIJH9w2C:-KN6c-va8k4MZL-MG8J6:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Source/OIDAuthorizationRequest.h#L176)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Same
</details>
Done.

---

*[Source/OIDAuthorizationRequest.h, line 178 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN39L-2q_Tz2WStD1x9:-KN6c10VBgsITYRKb3AI:-650086605) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Source/OIDAuthorizationRequest.h#L178)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Consider moving nullable parameters to end of list. No strong feeling if ordering makes more sense as is.
</details>
I  prefer it like this as I think that client ID and client secret are both credentials and should be one next to the other.

---

*[Source/macOS/OIDAuthorizationUICoordinatorMac.h, line 27 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KN3A1A5cMJxZa6IPmy9:-KN6cGPj_m-Ghe1BOPHb:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2e8a4cfa1238a74716656234fb8bd3e0810c3c0a/Source/macOS/OIDAuthorizationUICoordinatorMac.h#L27)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> In other files you've added a space between the base class and the protocol list.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/35)*
<!-- Sent from Reviewable.io -->

msarda commented:
Review status: 18 of 30 files reviewed at latest revision, 34 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KMOI3N3u4PAWv1A2Nbk:-KN70kQVndlZCLsO6Wtc:-2108161337) (no related file):*
<details><summary><i>Previously, msarda wrote…</i></summary>
> I fixed the test breakage. Please take a another look.
</details>
I addressed your comments. Note that when a comment on the macOS example applied to also to the existing iOS code, I've made the fix in the iOS code as well.

Please take another look.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/35)*
<!-- Sent from Reviewable.io -->

WilliamDenniss commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 16 of 26 files at r1, 2 of 2 files at r2, 12 of 12 files at r4.
Review status: all files reviewed at latest revision, 32 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/35#-:-KN7on3Zu2pJrUh6LFlj:1417088721)*
<!-- Sent from Reviewable.io -->

StevenEWright commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/openid/appauth-ios/35#-KMOI3N3u4PAWv1A2Nbk:-KN7xm-pF6rm9imFMYDR:1360895522) (no related file):*
<details><summary><i>Previously, msarda wrote…</i></summary>
> I addressed your comments. Note that when a comment on the macOS example applied to also to the existing iOS code, I've made the fix in the iOS code as well.
> 
> Please take another look.
</details>
:100: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/35#-:-KN7xBl7m2AX033vM-TI:1417088721)*
<!-- Sent from Reviewable.io -->
